src/main/cpp/src/datetime_parser.cu

```diff
@@ -0,0 +1,668 @@
+  // LIST<STRUCT<utcInstant: int64, tzInstant: int64, utcOffset: int32,
+  // looseTzInstant: int64>>.
+  thrust::optional<lists_column_device_view const> transitions = thrust::nullopt;
+  thrust::optional<column_device_view const> tz_indices        = thrust::nullopt;
+  thrust::optional<column_device_view const> tz_short_ids      = thrust::nullopt;
+
+  __device__ thrust::tuple<cudf::timestamp_us, uint8_t> operator()(const cudf::size_type& idx) const
+  {
+    // inherit the null mask of the input column
+    if (!d_strings.is_valid(idx)) {
+      return thrust::make_tuple(cudf::timestamp_us{cudf::duration_us{0}}, ParseResult::INVALID);
+    }
+
+    auto const d_str = d_strings.element<cudf::string_view>(idx);
+
+    timestamp_components ts_comp{};
+    char const* tz_lit_ptr = nullptr;
+    size_type tz_lit_len   = 0;
+    switch (parse_string_to_timestamp_us(&ts_comp, &tz_lit_ptr, &tz_lit_len, d_str)) {
+      case ParseResult::INVALID:
+        return thrust::make_tuple(cudf::timestamp_us{cudf::duration_us{0}}, ParseResult::INVALID);
+      case ParseResult::UNSUPPORTED:
+        return thrust::make_tuple(cudf::timestamp_us{cudf::duration_us{0}},
+                                  ParseResult::UNSUPPORTED);
+      case ParseResult::OK:
+      default: break;
+    }
+
+    if constexpr (!with_timezone) {
+      // path without timezone, in which unix_timestamp is straightforwardly
+      // computed
+      auto const ts_unaligned = compute_epoch_us(ts_comp);
+      return thrust::make_tuple(cudf::timestamp_us{cudf::duration_us{ts_unaligned}},
+                                ParseResult::OK);
+    }
+
+    // path with timezone, in which timezone offset has to be determined before
+    // computing unix_timestamp
+    int64_t tz_offset;
+    if (tz_lit_ptr == nullptr) {
+      tz_offset = extract_timezone_offset(compute_loose_epoch_s(ts_comp), default_tz_index);
+    } else {
+      auto tz_view = string_view(tz_lit_ptr, tz_lit_len);
+
+      // Map short TZ ID to region-based timezone if tz_view is a short ID
+      auto const& short_name_col   = tz_short_ids->child(0);
+      auto const& region_based_col = tz_short_ids->child(1);
+      for (size_type i = 0; i < tz_short_ids->size(); i++) {
+        auto const& curr_short_id = short_name_col.element<string_view>(i);
+        if (curr_short_id == tz_view) {
+          // find short ID, replace tz_view with mapped region TZ ID
+          tz_view = region_based_col.element<string_view>(i);
+          break;
+        }
+      }
+
+      // Firstly, try parsing as utc-like timezone rep
+      if (auto [utc_offset, ret_code] = parse_utc_like_tz(tz_view); ret_code == 0) {
+        tz_offset = utc_offset;
+      } else if (ret_code == 1) {
+        // Then, try parsing as region-based timezone ID
+        auto tz_index = query_index_from_tz_db(tz_view);
+        // tz_index < size(tzDB): found the ID in tzDB
+        // size(tzDB) <= tz_index < size(tzIDs): found the ID but not supported
```

Comment:
Why not make the invalid indexes negative? Should we have some high level APIs to deal with looking things up in the tz_db so it is not just by convention that we all know to look to see if it is out of bounds, and then have to know what that means.

Reply:
> Why not make the invalid indexes negative?

Done